File: app/src/main/java/com/ivan/github/account/UserCenterImpl.java
package com.ivan.github.account;

import android.content.SharedPreferences;
import android.text.TextUtils;

import com.ivan.github.GitHub;
import com.ivan.github.account.model.User;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

/**
 * UserCenter implementation
 */
public class UserCenterImpl implements IUserCenter, SharedPreferences.OnSharedPreferenceChangeListener {

    private static final String USER_DETAIL = "user_detail";
    private static final String AUTH_KEY = "auth_key";

    private static volatile UserCenterImpl mUserCenter;

    private User mUser;
    private String mAuthorization;
    private final Set<ILoginStatusChangedListener> mStatusChangedListeners;
    private final SharedPreferences mPreference;


    private UserCenterImpl() {
        mStatusChangedListeners = new CopyOnWriteArraySet<>();
        mPreference = GitHub.appComponent().secureSharedPreference();
        mPreference.registerOnSharedPreferenceChangeListener(this);
    }

    public static UserCenterImpl getInstance() {
        if (mUserCenter == null) {
            synchronized (UserCenterImpl.class) {
                if (mUserCenter == null) {
                    mUserCenter = new UserCenterImpl();
                }
            }
        }
        return mUserCenter;
    }

    public void init(User user, String auth) {
        this.mUser = user;
        this.mAuthorization = auth;
        GitHub.appComponent().secureSharedPreference().edit().putString(AUTH_KEY, auth).apply();
    }

    public void saveUser(User user) {
        mUser = user;
        if (mUser != null) {
            String userStr = GitHub.appComponent().gson().toJson(mUser);
            mPreference.edit()
                    .putString(USER_DETAIL, userStr)
                    .apply();
        } else {
            mPreference.edit()
                    .putString(USER_DETAIL, null)
                    .apply();
        }
    }

    public void loadUser() {
        String user = mPreference.getString(USER_DETAIL, null);
        if (!TextUtils.isEmpty(user)) {
            mAuthorization = mPreference.getString(AUTH_KEY, "");
            mUser = GitHub.appComponent().gson().fromJson(user, User.class);
        }
    }

    @Override
    public void logout() {
        mUser = null;
        mAuthorization = null;
        mPreference.edit().putString(USER_DETAIL, null).putString(AUTH_KEY, null).apply();
    }

    @Override
    public String getAuthorization() {
        if (TextUtils.isEmpty(mAuthorization)) {
            mAuthorization = mPreference.getString(AUTH_KEY, "");
        }
        return mAuthorization;
    }

    @Override
    public void registerLoginStatusChangedListener(ILoginStatusChangedListener listener) {
        if (listener != null) {
            mStatusChangedListeners.add(listener);
        }
    }

    @Override
    public void unregisterLoginStatusChangedListener(ILoginStatusChangedListener listener) {
        if (listener != null) {
            mStatusChangedListeners.remove(listener);
        }
    }

    private void notifyLogin() {
        for (ILoginStatusChangedListener listener : mStatusChangedListeners) {
            listener.onStatusChanged(mUser, STATUS_LOGIN);
        }
    }

    private void notifyLogout() {
        for (ILoginStatusChangedListener listener : mStatusChangedListeners) {
            listener.onStatusChanged(null, STSTUS_LOGOUT);
        }
    }

    @Override
    public boolean isLogin() {
        return mUser != null && mAuthorization != null;
    }

    @Override
    public User getUser() {
        return mUser;
    }

    @Override
    public String getUsername() {
        if (mUser == null) {
            return "";
        } else {
            return mUser.login;
        }
    }

    @Override
    public String getLogin() {
        if (mUser == null) {
            return "";
        } else {
            return mUser.login;
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (USER_DETAIL.equals(key)) {
            if (mUser == null) {
                notifyLogout();
            } else {
                notifyLogin();
            }
        }
    }
}

File: app/src/main/kotlin/com/github/account/UserCenterImpl.kt
package com.github.account

import android.text.TextUtils
import com.github.GitHub
import com.github.account.model.User
import com.github.app.GHStorage
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.map
import java.util.concurrent.CopyOnWriteArraySet
import javax.inject.Inject
import javax.inject.Singleton

/**
 * UserCenter implementation using Vault (via GHStorage)
 */
@Singleton
class UserCenterImpl @Inject constructor(
    private val ghStorage: GHStorage
) : IUserCenter {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val storage by lazy { ghStorage.getSecureStorage() }

    private var mUser: User? = null
    private var mAuthorization: String? = null
    private val mStatusChangedListeners = CopyOnWriteArraySet<ILoginStatusChangedListener>()

    init {
        // Load initial state and watch for changes
        scope.launch {
            loadUserInternal()
            
            // Watch for external changes
            storage.watch(USER_DETAIL, User::class.java).collectLatest { user ->
                if (user == null && mUser != null) {
                    mUser = null
                    mAuthorization = null
                    notifyLogout()
                } else if (user != null && mUser == null) {
                    mUser = user
                    notifyLogin()
                }
            }
        }
    }

    override fun registerLoginStatusChangedListener(listener: ILoginStatusChangedListener?) {
        listener?.let { mStatusChangedListeners.add(it) }
    }

    override fun unregisterLoginStatusChangedListener(listener: ILoginStatusChangedListener?) {
        listener?.let { mStatusChangedListeners.remove(it) }
    }

    override fun isLogin(): Boolean = mUser != null && !mAuthorization.isNullOrEmpty()

    override fun isLoginFlow(): Flow<Boolean> = storage.watch(USER_DETAIL, User::class.java)
        .map { it != null }

    override fun getUser(): User? = mUser

    override fun getUsername(): String = mUser?.login ?: ""

    override fun getLogin(): String = mUser?.login ?: ""

    override fun logout() {
        mUser = null
        mAuthorization = null
        scope.launch {
            storage.remove(USER_DETAIL)
            storage.remove(AUTH_KEY)
        }
        notifyLogout()
    }

    override fun getAuthorization(): String? {
        return mAuthorization
    }

    override fun init(user: User?, auth: String?) {
        this.mUser = user
        this.mAuthorization = auth
        scope.launch {
            storage.put(AUTH_KEY, auth ?: "")
            if (user != null) {
                storage.put(USER_DETAIL, user)
            }
        }
    }

    override fun saveUser(user: User?) {
        this.mUser = user
        scope.launch {
            if (user != null) {
                storage.put(USER_DETAIL, user)
            } else {
                storage.remove(USER_DETAIL)
            }
        }
    }

    /**
     * Called by UserCenterInit to load user data on startup.
     */
    fun loadUser() {
        scope.launch {
            loadUserInternal()
        }
    }

    private suspend fun loadUserInternal() {
        val userResult = storage.get(USER_DETAIL, User::class.java)
        mUser = userResult.getOrNull()
        
        val authResult = storage.get(AUTH_KEY, String::class.java)
        mAuthorization = authResult.getOrNull()
    }

    private fun notifyLogin() {
        mStatusChangedListeners.forEach { it.onStatusChanged(mUser, IUserCenter.STATUS_LOGIN) }
    }

    private fun notifyLogout() {
        mStatusChangedListeners.forEach { it.onStatusChanged(null, IUserCenter.STATUS_LOGOUT) }
    }

    companion object {
        private const val USER_DETAIL = "user_detail"
        private const val AUTH_KEY = "auth_key"

        @JvmStatic
        fun getInstance(): UserCenterImpl {
            // For legacy Java access
            return GitHub.appComponent().userCenter() as UserCenterImpl
        }
    }
}
